### ISSUE TITLE
Task: Implement Performance-Aware Data Allocation in Scheduler

### ISSUE THREAD

**Issue by orlandohohmeier** [hypha-space/hypha#79]
### Summary
Extend the scheduler to dynamically adjust data allocation among workers based on their observed processing performance.  
The scheduler should leverage the data slicing mechanism (from #64) and upcoming job progress tracking to distribute more data to faster workers, ensuring efficient synchronization across nodes and improved total throughput.

Key goals:
- Integrate with the new **data slicing and provisioning logic** (#64).  
- Use **job status tracking** to monitor worker performance (throughput, progress).  
- Adjust **data slice sizes** dynamically across training rounds:
  - Faster workers receive proportionally more data.
  - Slower workers receive less, reducing idle time for others.  
- Ensure all workers complete a round in _roughly the same time_, reducing bottlenecks.  
- Maintain data consistency and ensure each portion of the dataset is processed exactly once.  

Structure this as a **larger feature** that may require **multiple supporting tasks** to implement.

### Acceptance Criteria

- Given that workers have varying processing speeds,
  When scheduling new training rounds,
  Then the scheduler should allocate larger data slices to faster workers and smaller slices to slower ones.

- Given that job progress is tracked,
  When a worker completes its assigned slice,
  Then the scheduler should use that information to refine future allocations.

- Given that data allocation has been adapted,
  When the next training round starts,
  Then workers should progress at a more balanced pace, minimizing idle time at synchronization points.

### Background
Currently, the scheduler distributes data slices evenly across workers, assuming similar performance.  
However, in heterogeneous environments — where some nodes have significantly more compute or memory — this leads to inefficiencies: faster workers finish early and remain idle while slower ones complete their work.

This feature introduces **adaptive, performance-aware data allocation**, using metrics from job tracking to inform future scheduling decisions.  
By coupling the **data slicing mechanism (#64)** with **job status monitoring**, the scheduler can continuously rebalance workload distribution.  

The result is more synchronized progress across workers and improved overall training efficiency, especially in federated or decentralized training setups.

As this is a broader and more involved effort, it is expected to span multiple subtasks — such as integration with job tracking, modifications to data provisioning logic, and scheduler decision-making refinements.

**Comment by l45k** [comment]
For the batch scheduling for the worker, we should track the batch_size per worker, the remaining data points until the next update and the average time in `ms` a worker needs to process a batch. With that we can project into the future how many rounds each worker should work until the next parameter update. We can start with
```Python
def project(to_go: int, response_times: np.array, batch_sizes: np.array, time_steps: int = 10, time_cap: int = 5 * 60 * 1000):
  updates = np.zeros(batch_sizes.shape)
  next_update = deepcopy(response_times)
  time = 0
  done = 0
  while np.max(updates) < time_steps and time < time_cap and (to_go - done) > 0:
    update_idx = np.argmin(next_update)
    time = next_update[update_idx]
    done += batch_sizes[update_idx]
    updates[update_idx] += 1
    next_update[update_idx] += response_times[update_idx]
  return time, done, (to_go - done) , updates
```
as a baseline for the projection function.
Given the batch_sizes, average batch time and data points until the next update, it will return how many batches each worker should process. The projection can be limited by either a time cap (something like don't projekt longer than 5 minutes) or maximum of batches (not more than x batches on a single worker).
This can then be used to decide when will be a good time to schedule an update, thus informing the worker and the parameter server. This should provide the building blocks to data aware scheduling.
